nselib/punycode.lua

```diff
@@ -0,0 +1,664 @@
+-- @return Returns encoded/decoded string based on the formatter.
+-- The following function is adapted from punycode.js by Mathias Bynens
+-- under the MIT License.
+function mapLabels(labels, fn, formatter, delimiter)
+
+  local encoded = {}
+
+  for index, v in ipairs(labels) do
+
+    local res, err = fn(labels[index], formatter)
+
+    if err then
+      stdnse.debug2(err)
+      return nil
+    end
+
+    encoded[index] = res
+  end
+
+  return table.concat(encoded, delimiter)
+
+end
+
+---
+-- This function converts the given domain name or string into a
+-- ASCII string.
+--
+-- @param input Domain or string to be decoded.
+-- @param decoder A decoder function to convert the domain into a
+-- table of Unicode code points. Default: unicode.utf8_dec
+-- @param encoder An encoder function to convert a Unicode code
+-- point into a string of bytes.
+-- @param decoder An decoder function to decode the input string
+-- into an array of code points.
+-- @return Returns decoded string in the desired format.
+-- @return Throws an error, if any.
+function encode(input, encoder, decoder)
+
+  decoder = decoder or unicode.utf8_dec
+  encoder = encoder or unicode.utf8_enc
+
+  local decoded_tbl = unicode.decode(input, decoder)
+
+  -- Works only for punycode.
+  for _, val in pairs(regexSeparators) do
+    decoded_tbl = find_and_replace(decoded_tbl, val, byte('.'))
+  end
+
+  local delimiterCodePoint = 0x002E
+  -- Expects codepoints and delimiter values.
+  local codepointLabels = idna.breakInput(decoded_tbl, delimiterCodePoint)
+
+  local stringLabels = {}
+
+  for _, label in ipairs(codepointLabels) do
+    table.insert(stringLabels, unicode.encode(label, encoder))
+  end
+
+  local delimiter = unicode.encode({0x002E}, encoder)
+
+  return mapLabels(stringLabels, encode_label, decoder, delimiter)
+end
+
+---
+-- This function converts the given domain name or string into a
+-- Unicode string.
+--
+-- @param input Domain or string to be encoded.
+-- @param encoder An encoder function to convert a Unicode code
+-- point into a string of bytes.
+-- @param decoder An decoder function to decode the input string
+-- into an array of code points.
+-- @return Returns encoded string in the desired format.
+-- @return Throws an error, if any.
+function decode(input, encoder, decoder)
+
+  encoder = encoder or unicode.utf8_enc
+  decoder = decoder or unicode.utf8_dec
+  local delimiterCodePoint = 0x002E
+  local delimiter = unicode.encode({0x002E}, encoder)
+
+  codepoints = unicode.decode(input, decoder)
+  local codepointLabels = idna.breakInput(codepoints, delimiterCodePoint)
+
+  local stringLabels = {}
+
+  for _, label in ipairs(codepointLabels) do
+    table.insert(stringLabels, unicode.encode(label, encoder))
+  end
+
+  return mapLabels(stringLabels, decode_label, encoder, delimiter)
+
+end
+
+--Ignore the rest if we are not testing.
+if not unittest.testing() then
+  return _ENV
+end
+
+-- Table of punycode test cases.
+local testCases = {
```

Comment:
When you convert the libraries to move the domain splitting/joining and Unicode stuff to `idna.lua`, you'll have to change these test cases to each be a single label.

Reply:
Done.